### PR TITLE
Add support for creating topo geometry 

### PR DESCRIPTION
**Pull request by xylar**
In this mode, this requires:
* extracting edge polygons, not just cell polygons
* adding a `boundaryMask` field indicating parts of edge polygons
  with no neighboring cell
* extracting fields on cells as point fields, not cell fields in
  ParaView
* Sampling fields that are functions of depth at the topography
  index (e.g. `maxLevelCell`)

Python 3 support has been added.

PEP8 formatting has been fixed (because this was helpful in debugging Python 3 support).

### CODE REVIEW

**Comment by xylar** [comment]
@pwolfram, this should be enough to create the "water-tight" visualizations you were asking for.  Let me know if that's the case.  I'd like to have this reviewed and merged by the end of the year if at all possible because otherwise it might get forgotten.

**Comment by xylar** [comment]
I tested this with a QU240wLI test case.  I extracted temperature (as well as bottomDepth and maxLevelCell) and was able to produce this movie:

https://drive.google.com/open?id=1iAkGd-ZPbdQmhF3Eiusu8dldJwdsQjpH

**Comment by pwolfram** [comment]
I tested the new commit with 
```
python MPAS-Tools/visualization/paraview_vtk_field_extractor/paraview_vtk_field_extractor.py -f output/output.0001-01-01_00.00.00.nc --topo_dim nVertLevels --topo_cell_index maxLevelCell --lonlat -o tmp_topography -v ssh,layerThickness,bottomDepth -m input.nc --combine -t 0:10
```

with an issue using latlon coordinates:

image

But, with latlon off:

```
rm -rf tmp_topography/; python MPAS-Tools/visualization/paraview_vtk_field_extractor/paraview_vtk_field_extractor.py -f output/output.0001-01-01_00.00.00.nc --topo_dim nVertLevels --topo_cell_index maxLevelCell  -o tmp_topography -v ssh,layerThickness,bottomDepth -m input.nc --combine -t 0:10 
```

a nice output of 

image

using the ParaView calculator of `coords - coords/mag(coords)*bottomDepth*1000` with 'Coordinate Results' checked.

File: visualization/paraview_vtk_field_extractor/utils.py [hunk 1]
```diff
@@ -306,45 +307,67 @@ def parse_index_string(index_string, dim_size):  # {{{
...
+    if dimension_list is not None:
+        for dim_item in dimension_list:
+            (dimName, index_string) = dim_item.split('=')
+            indices = parse_extra_dim(dimName, index_string, time_series_file,
+                                      mesh_file)
+            if indices is not None:
+                if max_index_count is None or len(indices) <= max_index_count:
+                    extra_dims[dimName] = indices
+                else:
+                    extra_dims[dimName] = indices[0:max_index_count]
```
[the remark this answers is not in the document]
Comment (xylar) [reply visualization/paraview_vtk_field_extractor/utils.py]: Yeah, I'm typically explicit about the min if there's a max.

File: visualization/paraview_vtk_field_extractor/utils.py [hunk 2]
```diff
@@ -589,9 +627,131 @@ def write_vtp_header(path, prefix, active_var_index, var_indices,
     return vtkFile  # }}}
 
 
+def build_topo_point_and_polygon_lists(nc_file, output_32bit, lonlat):  # {{{
```
[the remark this answers is not in the document]
Comment (xylar) [reply visualization/paraview_vtk_field_extractor/utils.py]: Yeah, I do think so.  The other functions from which this one is derived weren't documented and I thought it was kind of a can of worms but I don't mind documenting this one and leaving the other ones undocumented for now.

File: visualization/paraview_vtk_field_extractor/paraview_vtk_field_extractor.py [hunk 3]
```diff
@@ -200,8 +245,8 @@ def build_field_time_series(local_time_indices, file_names, mesh_file,
                                      "{}".format(xtimeName, time_series_file))
                 var = time_series_file.variables[xtimeName]
                 if len(var.shape) == 2:
-                    xtime = \
-                        ''.join(var[local_time_indices[time_index], :]).strip()
+                    xtime = var[local_time_indices[time_index],
+                                :].tostring().decode('utf-8').strip()
```
[the remark this answers is not in the document]
Comment (xylar) [reply visualization/paraview_vtk_field_extractor/paraview_vtk_field_extractor.py]: The "decode" but is new so this isn't just PEP8.  It's python 3 support.  But, yeah, this change should be in the Python 3/PEP8 commit.  I agree that it's not entirely in the scope of the PR but I can't be bothered to have 2 separate PRs in this case...

[the remark this answers is not in the document]
Comment (pwolfram) [reply visualization/paraview_vtk_field_extractor/paraview_vtk_field_extractor.py]: I missed seeing `decode` on my first pass here.  Thanks for pointing that out.

**Comment by xylar** [comment]
@pwolfram, could you redo your plot from above (the one that's *not* lon/lat) with:
```
  coords * (1.0 - 1000*((1 - boundaryMask)*bottomDepth + 30*boundaryMask)/mag(coords))
```
changing the numbers (1000 -- the stretch and 30 -- the value for "sea level") to whatever makes sense?

That would check if things look more like you were hoping.

File: visualization/paraview_vtk_field_extractor/utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (pwolfram) [reply visualization/paraview_vtk_field_extractor/utils.py]: I don't think it is a big deal, I wouldn't worry about it until it becomes a bigger issue.  A lot of the smaller functions are helpers for the bigger functions that are documented.  Perhaps the solution is to indicate they are private via a `_` prefix?  But, this isn't a big deal and folks don't uniformly conform to this style.

[the remark this answers is not in the document]
Comment (pwolfram) [reply visualization/paraview_vtk_field_extractor/utils.py]: Let's not worry about this for now but keep it in mind if functionality extends past the scope we are dealing with already, e.g., we go past 1-3k lines of code.  I would say this is needed more for 10k lines of code sized project.

**Comment by xylar** [comment]
@pwolfram, I think I've fixed the periodicity issues (at least it worked for me in the case I tried).  So far, I only tested `--lonlat` and not with a planar, periodic mesh.  If you have one of the latter around that you could try, that might be useful.  I didn't have anything handy.

**Comment by xylar** [comment]
Sorry, something is still wrong at the periodic boundary.  The geometry is correct but the cell data seems to be coming from the wrong cell(s):

lonlat

**Comment by xylar** [comment]
... and fixed:

lonlat_fixed

Very satisfying!

**Comment by pwolfram** [comment]
Thanks @xylar, I'd say this PR fully meets the visualization goals we discussed over the last week or so.  I'm happy to do the merge or you can if you'd prefer.

**Comment by pwolfram** [comment]
Thanks @xylar, I'd say this PR fully meets the visualization goals we discussed over the last week or so.  I'm happy to do the merge or you can if you'd prefer.

**Comment by pwolfram** [comment]
Some notes of the transform in 2D for the review images:

for Coordinate Result of bathymetry
` coords - 3*((1 - boundaryMask)*bottomDepth -20.0*boundaryMask)/600.0*kHat`
for Coordinate Result of ssh
`  coords + 3*((1 - boundaryMask)*ssh -0.0*boundaryMask)/600.0*kHat`

Threshold on `wet=bottomDepth + ssh > 4e-3` to plot ssh.

**Comment by xylar** [comment]
Thanks @pwolfram, I'll merge.

**Pull request merged**